compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs

```diff
@@ -3365,4 +3365,113 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             err.span_label(block.span, "this block is missing a tail expression");
         }
     }
+
+    pub(crate) fn annotate_incorrect_or_expr(
+        &self,
+        diag: &mut Diag<'_>,
+        expr: &'tcx hir::Expr<'tcx>,
+        expr_ty: Ty<'tcx>,
+        expected_ty: Ty<'tcx>,
+    ) {
+        if expected_ty != self.tcx.types.bool {
+            return;
+        }
+        let hir::Node::Expr(&hir::Expr {
+            kind:
+                hir::ExprKind::Binary(
+                    hir::BinOp { node: hir::BinOpKind::Or, span: binop_span },
+                    lhs,
+                    rhs,
+                ),
+            hir_id: parent_hir_id,
+            span: full_span,
+            ..
+        }) = self.tcx.parent_hir_node(expr.hir_id)
+        else {
+            return;
+        };
+        if rhs.hir_id != expr.hir_id {
+            return;
+        }
+        let hir::Expr {
+            kind:
+                hir::ExprKind::Binary(hir::BinOp { node: hir::BinOpKind::Eq, span: eq_span }, lhs, _),
+            ..
+        } = *lhs
+        else {
+            return;
+        };
+        let Some(lhs_ty) = self.typeck_results.borrow().expr_ty_opt(lhs) else {
+            return;
+        };
+        // Coercion here is not totally right, but w/e.
+        if !self.can_coerce(expr_ty, lhs_ty) {
+            return;
+        }
+
+        // Track whether all of the exprs to the right, i.e. `|| a || b || c` are all pattern-like.
+        let mut is_literal = rhs.is_approximately_pattern();
+        // Track the span of the outermost `||` expr.
+        let mut full_span = full_span;
+
+        // Walk up the expr tree gathering up the binop spans of any subsequent `|| a || b || c`.
```

Comment:
You might want to move this comment about walking to before the comment for `is_literal` so the reader knows what we would be doing better